pkg/expr/identity_test.go

```diff
@@ -143,38 +145,73 @@ func TestEvalPredicate(t *testing.T) {
+		}, in: "badExpr1", out: "bar", err: "invalid expression"},
 		{attrs: &mixerpb.Attributes{
-			Dictionary:       dictionary{0: "bool", 1: "str"},
-			BoolAttributes:   map[int32]bool{0: true},
-			StringAttributes: map[int32]string{1: "foo"},
-		}, in: "bool", out: true, err: false},
-		{attrs: &mixerpb.Attributes{Dictionary: dictionary{5: "time"}, TimestampAttributes: map[int32]*ts.Timestamp{5: ts1}}, in: "time", out: nil, err: true},
-		{attrs: &mixerpb.Attributes{Dictionary: dictionary{2: "a"}, BoolAttributes: map[int32]bool{2: true}}, in: "b", out: false, err: true},
+			Dictionary:       dictionary{1: "foo", 2: "bar"},
+			StringAttributes: map[int32]string{1: "foo", 2: "bar"},
+		}, in: "badExpr2==", out: false, err: ""},
+		{attrs: &mixerpb.Attributes{
+			Dictionary:       dictionary{1: "foo", 2: "bar"},
+			StringAttributes: map[int32]string{1: "foo", 2: "bar"},
+		}, in: "$badExpr3==", out: "bar", err: "unresolved attribute"},
+		{attrs: &mixerpb.Attributes{
+			Dictionary:       dictionary{1: "foo", 2: "bar"},
+			StringAttributes: map[int32]string{1: "foo", 2: "bar"},
+		}, in: "==$badExpr3", out: "bar", err: "unresolved attribute"},
+		{attrs: &mixerpb.Attributes{
+			Dictionary:       dictionary{1: "foo", 2: "bar"},
+			StringAttributes: map[int32]string{1: "foo", 2: "bar"},
+		}, in: "==$", out: "bar", err: "empty attribute name"},
+		{attrs: &mixerpb.Attributes{
+			Dictionary:       dictionary{1: "foo", 2: "bar"},
+			StringAttributes: map[int32]string{1: "foo", 2: "bar"},
+		}, in: "$bar==bar", out: true, err: ""},
+		{attrs: &mixerpb.Attributes{
+			Dictionary:       dictionary{1: "foo", 2: "bar"},
+			StringAttributes: map[int32]string{1: "foo", 2: "bar"},
+		}, in: "$bar==foo", out: false, err: ""},
+		{attrs: &mixerpb.Attributes{
+			Dictionary:       dictionary{1: "foo", 2: "bar"},
+			StringAttributes: map[int32]string{1: "foo", 2: "bar"},
+		}, in: "$bar==foo", out: false, err: ""},
+		{attrs: &mixerpb.Attributes{
+			Dictionary:       dictionary{1: "foo", 2: "bar"},
+			StringAttributes: map[int32]string{1: "foo", 2: "bar"},
+		}, in: "$bar == $foo", out: false, err: ""},
+		{attrs: &mixerpb.Attributes{
+			Dictionary:       dictionary{1: "foo", 2: "bar"},
+			StringAttributes: map[int32]string{1: "foo", 2: "foo"},
+		}, in: "$bar == $foo", out: true, err: ""},
 	}
-	for _, c := range cases {
+	for idx, c := range cases {
 		bag, err := attribute.NewManager().NewTracker().StartRequest(c.attrs)
 		if err != nil {
 			t.Errorf("Failed to create bag for case: %v; err: %v", c, err)
 		}
 		id := NewIdentityEvaluator()
+		if err = id.Validate(c.in); err != nil {
+			t.Errorf("[%d] Unexpected error %s", idx, err.Error())
+		}
 		val, err := id.EvalPredicate(c.in, bag)
-		if c.err && err == nil {
-			t.Errorf("Expected err on case '%+v' but got none.", c)
+		if c.err != "" && err == nil {
```

Comment:
i think this might be clearer with some slight organization changes. something like:

```
if c.err == "" {
    if err != nil {
       t.Errorf("unexpected error...")
       continue
    }
    if val != c.out {
       t.Errorf("got %v, want %v"...)
    }
    continue
}
if err == nil {
    t.Errorf("expected error"...)
    continue
}
if !strings.Contains(err.Error(), c.err) {
    t.Errorf("got %v, want %v"...)
}
```

Following the `c.err ==` and `c.err !=` as they weave is a bit tricky.